[Video] Improve episode display/selection for multi-episode discs.

78andyp opened the pull request:
This is an evolution of #24178, which I have closed as I have made many changes/improvements.

I've been looking at how episodes are played when they are in a blu-ray ISO

EDIT: Bluray scraping now fixed in v21/master
EDIT: Simple DVD menu now supported (like bluray) - this allows selection of episode without DVD menu.

Currently you are asked to select from a list of titles - these are presented in no particular order and (by default) only show those with length 70%+ of the longest.

So, for example, on The Last of Us (UHD Disc 1) - if I try and play episode 2 (disc 1 contains episodes 1 and 2), I get the following:
<img width="452" alt="before" src="https://github.com/xbmc/xbmc/assets/99039295/6db5e22f-c431-4fe5-b4a3-426b98c6be20">

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

This demonstrates a number of issues:
1) No clear idea what titles 801 and 851 contain - they actually are both episode 1 (despite us asking to play episode 2)
2) No idea what languages are associated with each title (801 = episode 1 european languages, 851 = episode 1 japanese)

I've updated the code - so now when I select episode 2 I get the following:
<img width="804" alt="image" src="https://github.com/xbmc/xbmc/assets/99039295/e10cc6cf-7b95-4e49-82f0-6d8cb54170ec">

This also makes the Videos -> Files -> ... more logical as well.

Instead of:
![Before](https://github.com/xbmc/xbmc/assets/99039295/fbfcd807-71fc-4213-b84f-451e9e8c62fa)

We now have:
<img width="1280" alt="image" src="https://github.com/xbmc/xbmc/assets/99039295/4882ef7e-66b6-4a92-9f30-b66729195326">

Showing all the episodes on the disc instead of just the first one (with no indication the others are there)

and:
<img width="803" alt="image" src="https://github.com/xbmc/xbmc/assets/99039295/765d7425-4bcc-445a-9a87-60d94a4aba77">

Allowing you to sellect each episode rather than a random title list.

Methodology (from code comments)

```
  // At this stage we have a number of ways of trying to determine the correct playlist for an episode
  //
  // 1) Using a 'Play All' playlist
  //    - Caveat: there may not be one
  //
  // 2) Using the longest (non-'Play All') playlists that are consecutive
  //    - Caveat: assumes play-all detection has worked and playlists are consecutive (hopefully reasonable assumptions)
  //
  // 3) Playlists that are +/- 30sec of the episode length that is passed to us
  //    - Caveat: the episode length may be wrong - either from scraper/NFO or from bug in Kodi that overwrites episode streamdetails on same disc
  //
  // 4) Using position in a group of adjacent playlists (see below)
  //    - Cavet: won't work for single episode discs
  // 
  // 5) - For single episode discs - this is hard. There are some discs where there are extras that are longer than the episode itself
  //      The only obvious difference sometimes seems to be number of languages
  //
  // Order of preference:
  //
  // 1) Use 'Play All' playlist - take the nth clip and find a playlist that plays that clip
  //
  // 2) Take the nth playlist of a consecutive run of longest playlists
  //
  // 3) Refine a group using relative position of playlist found on basis of length
  //    - this tries to account for episodes that have the wrong duration passed
  //    eg. if we have a disc with episodes 4,5,6 we would expect episode 5 to be the middle playlist of a group of 3 consecutive playlists (eg. 801, 802, 803)
  //
  // 4) Playlist based on episode length alone (assumes no groups found or 2) and 3) failed)
  //
  // 5) Look at groups found (ignoring length) and pick the relevant playlist - eg. the middle one for episode 5 using the exampe above
  //    - Only looks at playlists > MIN_EPISODE_LENGTH
  // 
  // 6) For single episode discs only, look at the longest playlist with multiple languages
  //
  // SPECIALS
  //
  // These are more difficult - as there may only be one per disc and we can't make assumptions about playlists.
  // So have to look on basis of duration alone.
  //
```

## Motivation and context

Improve the rather unintuitive episode selection when episodes on bluray.

## How has this been tested?

Windows 11 x64

## What is the effect on users?

More intuitve episode selection when on bluray.

## Screenshots (if appropriate):

See above

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

78andyp commented:
@fuzzard - is there something wrong with this for Omega??

fuzzard commented:
The only reason we have gone to beta 3 is because of the version crap. Anything that touches db, or has any potential risk I'm pushing to 22, as team members are spending all their time trying to clean up after the version stuff.

78andyp commented:
> The only reason we have gone to beta 3 is because of the version crap. Anything that touches db, or has any potential risk I'm pushing to 22, as team members are spending all their time trying to clean up after the version stuff.

Thanks. Fair enough. This doesn't touch db btw... thats #24277. Anything I can do to help with versions, as it links with the ISO stuff i've been looking at??

fuzzard commented:
@ksooo would be the best person for guidance. 

From what I understand, the focus is getting the UI shit into a non terrible state that will hopefully not require a rework in 22. Then all the backend shit will need a massive overhaul for 22.

ksooo commented:
> @ksooo would be the best person for guidance.

As far as I can see this is not related to the video versions/extras feature, BUT as we are nearing a release I'm kind of hesitant to include such invasive (thus risky) change into Omega. Originally, we at this time wanted to be far beyond adding features for Omega, just because we now have to bring video versions/extras into good shape, that does not mean that we're open for adding any features to Omega again.

Videos/extras fixes is an exception with respect to risk, all other features/improvements should be postponed to v22.

78andyp committed:
[Video] Improve episode display/selection for multi-episode discs.

	modified:   addons/resource.language.en_gb/resources/strings.po
	modified:   system/settings/settings.xml
	modified:   xbmc/FileItem.cpp
	modified:   xbmc/FileItem.h
	modified:   xbmc/application/Application.cpp
	modified:   xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp
	modified:   xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamBluray.cpp
	modified:   xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.cpp
	modified:   xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamNavigator.h
	modified:   xbmc/dialogs/GUIDialogSimpleMenu.cpp
	modified:   xbmc/dialogs/GUIDialogSimpleMenu.h
	modified:   xbmc/filesystem/BlurayDirectory.cpp
	modified:   xbmc/filesystem/BlurayDirectory.h
	modified:   xbmc/filesystem/CMakeLists.txt
	modified:   xbmc/filesystem/DVDDirectory.cpp
	modified:   xbmc/filesystem/DVDDirectory.h
	new file:   xbmc/filesystem/DVDFile.cpp
	new file:   xbmc/filesystem/DVDFile.h
	modified:   xbmc/filesystem/Directory.cpp
	modified:   xbmc/filesystem/Directory.h
	modified:   xbmc/filesystem/FileFactory.cpp
	modified:   xbmc/filesystem/IDirectory.h
	modified:   xbmc/settings/DiscSettings.h
	modified:   xbmc/settings/Settings.h
	modified:   xbmc/utils/SaveFileStateJob.cpp
	modified:   xbmc/utils/URIUtils.cpp
	modified:   xbmc/video/VideoDatabase.cpp
	modified:   xbmc/video/VideoDatabase.h
	modified:   xbmc/video/guilib/VideoGUIUtils.cpp
	modified:   xbmc/video/windows/GUIWindowVideoBase.cpp

78andyp committed:
Rebase fixes
